specific/cmo_sale/models/sale.py

```diff
@@ -147,19 +152,25 @@ def cal_management_fee(self):
         }
 
     @api.multi
-    @api.onchange('price_unit', 'purchase_price')
+    @api.depends('sale_order_line_margin', 'price_unit',
+                 'purchase_price', 'product_uom_qty',
+                 )
+    @api.onchange('price_unit', 'purchase_price', 'product_uom_qty')
     def _compute_sale_order_line_margin(self):
         for line in self:
-            margin = line.price_unit - line.purchase_price
+            margin = (line.price_unit - line.purchase_price) * \
+                line.product_uom_qty
             line.sale_order_line_margin = margin
 
     @api.multi
+    @api.depends('so_line_percent_margin', 'price_unit', 'purchase_price')
     @api.onchange('price_unit', 'purchase_price')
     def _compute_so_line_percent_margin(self):
         for line in self:
-            margin = line.price_unit - line.purchase_price
+            margin = (line.price_unit - line.purchase_price)
             if line.price_unit:
-                line.so_line_percent_margin = margin * 100.0 / line.price_unit
```

Comment:
This is already correct. We have tested the line.price_unit already. So we don't need to test it again.